server/src/main/java/org/elasticsearch/action/fieldcaps/RequestDispatcher.java

```diff
@@ -191,67 +194,111 @@ private void sendRequestToNode(String nodeId, List<ShardId> shardIds) {
-            final IndexSelector indexSelector = indexSelectors.get(unmatchedShardId.getIndexName());
-            if (indexSelector != null) {
-                indexSelector.addUnmatchedShardId(unmatchedShardId);
+            final Group group = groups.get(unmatchedShardId.getIndexName());
+            if (group != null) {
+                group.addUnmatchedShardId(unmatchedShardId);
             }
         }
         for (Map.Entry<ShardId, Exception> e : nodeResponse.getFailures().entrySet()) {
-            final IndexSelector indexSelector = indexSelectors.get(e.getKey().getIndexName());
-            if (indexSelector != null) {
-                indexSelector.setFailure(e.getKey(), e.getValue());
+            final Group group = groups.get(e.getKey().getIndexName());
+            if (group != null) {
+                group.setFailure(e.getKey(), e.getValue());
             }
         }
-        afterRequestsCompleted(shardIds.size());
+        afterRequestsCompleted();
     }
 
     private void onRequestFailure(List<ShardId> shardIds, Exception e) {
         for (ShardId shardId : shardIds) {
-            final IndexSelector indexSelector = indexSelectors.get(shardId.getIndexName());
-            if (indexSelector != null) {
-                indexSelector.setFailure(shardId, e);
+            final Group group = groups.get(shardId.getIndexName());
+            if (group != null) {
+                group.setFailure(shardId, e);
             }
         }
-        afterRequestsCompleted(shardIds.size());
+        afterRequestsCompleted();
+    }
+
+    private List<Group> groupIndicesByMappingHash(
+        ClusterService clusterService,
+        ClusterState clusterState,
+        boolean withIndexFilter,
+        String[] indices
+    ) {
+        final Map<String, List<String>> withMappingHashes = new HashMap<>();
+        final List<Group> groups = new ArrayList<>();
+        for (String index : indices) {
+            final IndexMetadata indexMetadata = clusterState.metadata().index(index);
+            if (withIndexFilter == false
+                && indexMetadata != null
+                && indexMetadata.mapping() != null
+                && indexMetadata.mapping().getSha256() != null) {
+                withMappingHashes.computeIfAbsent(indexMetadata.mapping().getSha256(), k -> new ArrayList<>()).add(index);
+            } else {
+                final GroupShardsIterator<ShardIterator> shardIts = clusterService.operationRouting()
+                    .searchShards(clusterState, new String[] { index }, null, null, null, null);
+                groups.add(new Group(List.of(index), null, shardIts));
+            }
+        }
+        for (Map.Entry<String, List<String>> e : withMappingHashes.entrySet()) {
+            final GroupShardsIterator<ShardIterator> shardIts = clusterService.operationRouting()
+                .searchShards(clusterState, e.getValue().toArray(String[]::new), null, null, null, null);
```

Comment:
question: if the mapping is the same for some indices, can we just ask field_caps for a single index and apply what we get to all the other indices with the same hash?